extension/accounting/tracking.js

```diff
@@ -0,0 +1,268 @@
+    let endBucket = Math.floor(end / this.bucketInterval);
+    if(endBucket - beginBucket > this.maxBuckets || endBucket < beginBucket) {
+      log.info("Invalid bucket setup, skipped:", beginBucket, endBucket);
+      return [];
+    }
+    return [beginBucket, endBucket];
+  }
+  
+  async appendDestination(key, destination) {
+    const size = await rclient.scardAsync(key);
+    if (size >= this.maxItemsInBucket) {
+      return;
+    }
+    
+    await rclient.saddAsync(key, destination);
+    await rclient.expireAsync(key, this.expireInterval);
+  }
+  
+  async appendTraffic(key, traffic) {
+    await rclient.incrbyAsync(key, Math.floor(traffic));
+    await rclient.expireAsync(key, this.expireInterval);    
+  }
+  
+  // begin, end - timestamps
+  async recordDestination(mac, destination, begin, end) {
+    const buckets = this.getBuckets(begin, end);
+    if (buckets.length !== 2) {
+      return;
+    }
+    
+    for (let b = buckets[0]; b <= buckets[1]; b++) {
+      const key = this.getDestinationKey(mac, b);
+      await this.appendDestination(key, destination);
+    }      
+  }
+  
+  async recordTraffic(mac, flow, begin, end) {
+    const buckets = this.getBuckets(begin, end);
+    if (buckets.length !== 2) {
+      return;
+    }
+    
+    for (let b = buckets[0]; b <= buckets[1]; b++) {
+      const trafficKey = this.getTrafficKey(mac, b);
+      const duration = flow.du;
+      const traffic = flow.ob + flow.rb;
+      
+      // FIXME: may need to be more accurate
+      if (duration < this.bucketInterval / 1000) {
+        await this.appendTraffic(trafficKey, traffic);
+      } else {
+        await this.appendTraffic(trafficKey, traffic * this.bucketInterval / 1000 / duration);
+      }
+    }      
+  }
+  
+  async recordFlows(mac, flows) {
+    for(const flow of flows) {
+      const destIP = flowTool.getDestIP(flow);
+      const intel = await intelTool.getIntel(destIP);
+      if(intel && intel.b) { // ignore background traffic
+        continue;
+      }
+      const begin = flow.ts * 1000;
+      const end = flow.ets * 1000;
+      await this.recordDestination(mac, destIP, begin, end);
+      await this.recordTraffic(mac, flow, begin, end);
+    }
+  }
+  
+  // aggr the last 24 hours data, this is only for debugging purpose.
+  async aggrLast24Hours(mac) {
+    return this._aggr(mac, 24 * 3600 * 1000);
+  }
+  
+  async aggr(mac) {
+    return this._aggr(mac, this.bucketCountPerAggr * this.bucketInterval);
+  }
+  
+  async _aggr(mac, interval) {
+    const buckets = this.getBuckets(new Date() - interval, new Date());
+    if (buckets.length !== 2) {
+      return;
+    }
+    
+    let results = {};
+    
+    // traffic
+    const aggrTrafficKey = this.getAggregateTrafficKey(mac);
+    for(let b = buckets[0]; b <= buckets[1]; b++) {
+      const key = this.getTrafficKey(mac, b);
+      const x = await rclient.getAsync(key) || 0;
+      await rclient.hset(aggrTrafficKey, b, x);
```

Comment:
traffic & count aggregation looks like intermediate results. is saving necessary here?